### PR TITLE
MonadFail instances to allow building also with base >= 4.13

### PR DESCRIPTION
**Pull request by juhp**
I tested this allows building hawk from lts8 to lts15 (and nightly).

### CODE REVIEW

**Comment by gelisam** [comment]
Thanks!

**Comment by gelisam** [comment]
Wait, these CI failures look like genuine behavior changes, it's not the usual bitrot of the testsuite (which I have "fixed" by commenting out all the failing tests). Your changes don't look like they should change anything, but they do. Maybe some of the `fail` calls are now delegating to a `fail` implementation from a different layer in the Monad stack, thus causing e.g. IO to throw an IO exception instead of UncertainIO to propagate up a Left?

**Comment by gelisam** [comment]
... but looking at your changes again, I don't see how they can possibly change that. Maybe there are existing instances, e.g. ExceptT's, for which `Monad.fail` and `MonadFail.fail` have different behaviors??

File: haskell-awk.cabal [hunk 1]
```diff
@@ -38,7 +38,7 @@ library
       runtime
   ghc-options: -Wall
   build-depends:
-      base >=4.9.1.0 && <4.12.0.0
+      base >=4.9.1.0 && <5
```
Comment (gelisam) [inline haskell-awk.cabal:41]: as it says at the top, this file is generated by `package.yaml`, so I would have expected to see a corresponding change to that file

**Comment by gelisam** [comment]
Also, lts7? Wow, that's lower than https://github.com/gelisam/hawk/blob/master/stack-oldest-supported-lts.yaml, which means that hawk somehow became compatible with older versions of its dependencies than it previously was, that doesn't happen often! I don't see how adding support to MonadFail can have caused that to happen, so presumably some previous change did that and I didn't notice. This will require adjusting the lts in .travis.yml, stack-oldest-supported-lts.yaml and stack.yaml, and the lower bounds in package.yaml; but let's do that later, maybe in a separate PR, so we can focus on the current CI failures.

**Comment by juhp** [comment]
Sorry I think I meant from lts8, my bad :)

**Comment by juhp** [comment]
Hmm I see I should have run the testsuite before submitting this quick hack to get it to compile.

File: haskell-awk.cabal (hunk 1 above)
Comment (juhp) [reply haskell-awk.cabal:41]: Oops, I have fixed this easy part.

**Comment by juhp** [comment]
I changed it now to use `fail` instead of `MonadFail.fail`, let's see if that helps.

**Comment by gelisam** [comment]
I figured it out: the default implementation of `Monad.fail` is `error`, not `MonadFail.fail`, so you have to explicitly define `fail = MonadFail.fail` in every Monad instance. That's pretty dumb, especially since the MonadFail proposal intends to eventually remove the `Monad.fail` method. Oh well.

**Comment by juhp** [comment]
> I figured it out: the default implementation of `Monad.fail` is `error`, not `MonadFail.fail`, so you have to explicitly define `fail = MonadFail.fail` in every Monad instance. That's pretty dumb, especially since the MonadFail proposal intends to eventually remove the `Monad.fail` method. Oh well.

Ah great, well not great, but you know what I mean.. :)

**Comment by juhp** [comment]
I finally came back to this... (hope you weren't waiting. :-)

So funnily the current PR (a98afb7) actually passes for ghc-8.8, where `fail` finally moved fully from `Monad` to `MonadFail`).
(They would pass too for 8.10 probably I think, but for the format of some of the error message having changed there.)

https://downloads.haskell.org/~ghc/8.8.4/docs/html/users_guide/8.8.1-notes.html#compiler
https://wiki.haskell.org/MonadFail_Proposal

Your current code (dbf11a6) almost passes for ghc-8.6 if one adds `extensions: NoMonadFailDesugaring`.

**Comment by juhp** [comment]
I was thinking the easiest (only?) way to fix this and keep backward compatibility is to use CPP.
@gelisam are you okay with that?

**Comment by gelisam** [comment]
I don't think that's necessary unless you want to support a very wide range of ghc versions. The MonadFail proposal was very careful to introduce the change very slowly, so that it is possible to write code which is compatible with the last 3 ghc releases or something, which seems quite sufficient to me.

**Comment by juhp** [comment]
> I don't think that's necessary unless you want to support a very wide range of ghc versions. The MonadFail proposal was very careful to introduce the change very slowly, so that it is possible to write code which is compatible with the last 3 ghc releases or something, which seems quite sufficient to me.

Can you sketch how to do it then because I don't see it.

I suppose I could try to track down examples how other packages handled it - I remember there was a lot of breakage.

**Comment by gelisam** [comment]
[Here is the MonadFail proposal's rollout plan](https://wiki.haskell.org/MonadFail_Proposal#Transitional_strategy). As you can see, we are now at the final stage of the plan, when migrating to MonadFail is no longer optional. Because they took such a slow path to reach this stage, MonadFail has been available since 8.0, which is quite a few versions ago.

If we just switch from Monad.fail to MonadFail.fail without worrying about compatibility, we should be compatible with ghc 8.6+, and if we turn on `-XMonadFailDesugaring` (conditionally, since it no longer exists in ghc 8.8), then we can be compatible all the way to ghc 8.0+.

**Comment by gelisam** [comment]
Heads up: this is now blocking another volunteer (#237), so I would like to get this merged relatively soon (say, within a month). Would you like to get this over the finishing line, or should I? A third possibility which I have enjoyed recently is remote pair programming on open source projects, let me know if that sounds fun to you too!

**Comment by juhp** [comment]
Hi Samuel, thanks for the update.
Pair programming sounds fun - would be happy to do that if we can find a suitable time.
I certainly don't want to hold this back, so failing that also quite happy for you to go ahead with it. :-)

**Comment by gelisam** [comment]
Oops, I was clearly over-optimistic about the "within one month" part!

I see you're in Singapore. Since I'm in Montreal, finding a suitable time does seem like a challenge. According to [this chart](https://www.timeanddate.com/worldclock/meetingtime.html?iso=20210221&p1=165&p2=236), the least problematic time slots seems to be around:

1. Friday 9pm for me / Saturday 10am for you
2. Saturday 9pm for me / Sunday 10am for you

Would either work for you? If so, let's put something on the calendar so I don't again forget about our conversation for an entire month :)

**Comment by juhp** [comment]
Thanks, I sent you a private message - let me know if you don't see it.

**Comment by gelisam** [comment]
Closing in favor of #250

**Pull request closed without merging**